source/common/http/async_client_impl.cc

```diff
@@ -158,10 +158,8 @@ void AsyncStreamImpl::sendData(Buffer::Instance& data, bool end_stream) {
     return;
   }
 
-  // TODO(mattklein123): We trust callers currently to not do anything insane here if they set up
-  // buffering on an async client call. We should potentially think about limiting the size of
-  // buffering that we allow here.
-  if (buffered_body_ != nullptr) {
+  if (buffered_body_ != nullptr &&
+      buffered_body_->length() + data.length() <= buffered_body_size_limit_) {
```

Comment:
In the case where we exceed the limit, shouldn't we fail in some way? This is going to lead to corruption and hard to debug issues?

Reply:
If an attempt is made to perform a retry when the buffer is exceeded, a failure flag is added as follows:

```cpp
void AsyncStreamImpl::sendData(Buffer::Instance& data, bool) {
...
  if (buffer != nullptr) {
    if (isExceeded(buffer.size() + data.size(), BUFFER_LIMIT)) {
      should_fail_retry_ = true;
    } else {
      buffer.add(...)
    }
  }
...
}
```

If `should_fail_retry_` is set to `true`, it tries to delete the stream before scheduling the retry.
This method requires some interface changes, but it avoids implementations that delete the stream when the buffer is exceeded even though retries will never occur in the future. What do you think?

Reply:
For now, the current changes will fix the memory leak bug. Anyway, this change should be upstreamed immediately, and the follow-up changes can be implemented later (before the next release). In the meantime, we can deal with it by simply dumping the logs.

Reply:
Please add relevant TODOs, etc. You might also consider a power of 2 backoff warning log of some type until there is a better fix here. People should know this is happening.